File: Python/levelOrderRow.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: levelOrder.py
@Function: 从上到下打印二叉树
@Link: https://leetcode-cn.com/problems/cong-shang-dao-xia-da-yin-er-cha-shu-lcof/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-09-07
"""

from typing import List
import collections


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        res, queue = [], collections.deque()
        queue.append(root)
        while queue:
            tmp = []
            for _ in range(len(queue)):
                node = queue.popleft()
                tmp.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            res.append(tmp)
        return res


# ------------------------------
if __name__ == "__main__":
    # test on leetcode online.
    pass

File: C++/levelOrderRow.cpp
/**
 * @File    : levelOrderRow.cpp
 * @Brief   : 从上到下打印二叉树 II
 * @Link    : https://leetcode-cn.com/problems/cong-shang-dao-xia-da-yin-er-cha-shu-ii-lcof/
 * @Author  : Wei Li
 * @Date    : 2021-09-07
*/

#include <iostream>
#include <vector>
#include <queue>

// Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution
{
public:
    std::vector<std::vector<int>> levelOrder(TreeNode *root)
    {
        if (!root)
            return std::vector<std::vector<int>>();
        std::queue<TreeNode *> myqueue;
        std::vector<std::vector<int>> ans;

        myqueue.push(root);
        while (!myqueue.empty())
        {
            std::vector<int> nums;
            for (int count = myqueue.size(); count > 0; count--)
            {
                TreeNode *top = myqueue.front();
                nums.push_back(top->val);
                if (top->left)
                    myqueue.push(top->left);
                if (top->right)
                    myqueue.push(top->right);
                myqueue.pop();
            }
            ans.push_back(move(nums));
        }
        return ans;
    }
};

// -------------------------------
int main(int argc, char **argv)
{
    // test on leetcode online.

    return 0;
}
